src/aap_eda/services/ruleset/ruleset_manager.py

```diff
@@ -0,0 +1,257 @@
+        activation_db_logger: ActivationDbLogger,
+    ):
+        try:
+            activation = instance.activation
+            instance.status = ActivationStatus.FAILED
+            restart_policy = (
+                activation.restart_policy == RestartPolicy.ALWAYS.value
+                or activation.restart_policy == RestartPolicy.ON_FAILURE.value
+            )
+            restart_limit = activation.failure_count < int(
+                settings.ACTIVATION_MAX_RESTARTS_ON_FAILURE
+            )
+            if (
+                activation.is_enabled
+                and activation.is_valid
+                and restart_policy
+                and restart_limit
+            ):
+                self._log_restart_activation(
+                    error,
+                    activation,
+                    activation_db_logger,
+                    activation.failure_count + 1,
+                    int(settings.ACTIVATION_MAX_RESTARTS_ON_FAILURE),
+                )
+                activation.failure_count += 1
+                activation.save(update_fields=["failure_count", "modified_at"])
+            else:
+                more_reason = "unknown"
+                if not restart_limit:
+                    more_reason = (
+                        "it has exceeds the maximum number of restarts"
+                    )
+                elif not restart_policy:
+                    more_reason = "the restart policy is never"
+                elif not activation.is_valid:
+                    more_reason = "it is not restartable"
+                msg = (
+                    f"Activation {activation.name} failed: {str(error)}. "
+                    f"Will not restart because {more_reason}"
+                )
+                logger.error(msg)
+                activation_db_logger.write(msg)
+        except (IntegrityError, DatabaseError):
+            message = f"Failed to update instance [id: {instance.id}]"
+            logger.error(message)
+            raise ActivationRecordNotFound(message)
+
+    def _log_restart_activation(
+        self,
+        error: Exception,
+        activation: models.Activation,
+        activation_db_logger: ActivationDbLogger,
+        retry_count: int = 0,
+        max_retries: int = 0,
+    ) -> None:
+        if error:
+            seconds = int(settings.ACTIVATION_RESTART_SECONDS_ON_FAILURE)
+            msg = (
+                f"Activation {activation.name} failed: {str(error)}, "
+                f"retry ({retry_count}/{max_retries}) in {seconds} seconds "
+                "according to the activation's restart policy."
+            )
+            logger.warning(msg)
+        else:
+            seconds = int(settings.ACTIVATION_RESTART_SECONDS_ON_COMPLETE)
+            msg = (
+                f"Activation {activation.name} completed successfully. Will "
+                f"restart in {seconds} seconds according to its restart policy"
+            )
+            logger.info(msg)
```

Comment:
@hsong-rh Should this also have a command line generator method which can be called from the k8s and podman code, the cmdline is the same and we might want to add extra parameters down the road as we add new options into the Activation

Reply:
We already have something like that in the e2e test suite for rulebook: https://github.com/ansible/ansible-rulebook/blob/main/tests/e2e/utils.py#L18